File: import_build/src/AmbientSpace/Components/Geometry.py
import numpy as np

from src.Computation.State import State
from src.Computation.dState import dState


# A class used to store information about geometry of space
class Geometry:
    """
    A class used to store information about geometry of space

    ...

    Attributes
    ----------
    metricTensor : function
        function to generate 2D matrix representation of metric tensor at given position

    christoffel : function
        function to generate the terms corresponding to christoffel terms for acceleration of each vertex

    distance : function
        function to calculate the distance between two points in the ambient space

    Methods
    -------
    covariantAcceleration(state)
        Takes covariant derivative of state
        Returns dState storing initial velocity and covariant coordinates (second derivatives of coordinates) of the covariant derivatives

    dot(state1, state2)
        Calculates the dot product of velocity vectors for specificed states using the metric tensor
        Returns scalar value of dot product

    tangentBasis(pos)
        Calculates a basis for the tangent space at a point (in coordinates)
        Returns array of basis vectors

    gradient(fn, pos)
        Calculate the gradient of function fn at position pos
        Returns differential (gradient) of function at given point
    """

    def __init__(self, metricTensor, christoffel, distance):
        self.metricTensor = metricTensor
        self.christoffel  = christoffel
        self.distance     = distance


    def covariantAcceleration(self, state):
        vel = state.vel
        acc = self.christoffel(state)
        return dState(vel,acc)


    def dot(self, state1, state2):
        mat = self.metricTensor(state1.pos)
        v1 = state1.vel.copy()
        v2 = state2.vel.copy()
        # Apply this to the second vector
        gv2 = mat @ v2
        # Compute the dot product
        return v1.dot(gv2)


    # Get a basis for the tangent space at a point ( in coordinates )
    # @@@@ Consider changing how we want to do this, depending on which implementation of the gradient we want below:
    # @@@@ right now, this is the coordinate basis, and we use the metric tensor
    # @@@@ in the gradient: could instead do Gram-Schmidt here then calculate
    # @@@@ gradient as differential directly in that basis.
    def tangentBasis(self,pos):
        b1 = State(pos, np.array([1,0,0]))
        b2 = State(pos, np.array([0,1,0]))
        b3 = State(pos, np.array([0,0,1]))
        return np.array([b1,b2,b3])


    # //WARNING: IF THE COORDINATE SYSTEM IS SINGULAR: THIS COMPUTATION IS BAD AT THAT POINT!
    # //NEED GOOD COORDINATES.....
    # //get the gradient of a function fn at a position pos
    def gradient(self, fn, pos):

        basis = self.tangentBasis(pos)
        differential = State(pos, np.array([0,0,0]))

        #//add them all up:
        df0 = basis[0].differentiate(fn)
        b0 = basis[0].clone().multiplyScalar(df0)
        differential.add(b0)

        df1 = basis[1].differentiate(fn)
        b1 = basis[1].clone().multiplyScalar(df1)
        differential.add(b1)

        df2 = basis[2].differentiate(fn)
        b2 = basis[2].clone().multiplyScalar(df2)
        differential.add(b2)

        #//now the differential needs to be converted from a covector to a vector
        #//using the inverse metric:
        metric = self.metricTensor(pos)
        if(abs(np.linalg.det(metric))<0.00001):
            print('Warning! Metric Tensor Near Singular')
            print(pos)
            print(metric)

        invMetric = np.linalg.inv(metric.copy())
        differential.vel = invMetric @ differential.vel.copy()

        return differential

File: cpp_build/AmbientSpace/Components/Geometry.cpp
#include <vector>
#include <iostream>
#include <typeinfo>
#include <variant>
#include "../../includes/eigen-3.4.0/Eigen/Dense"
#include "Geometry.h"
#include "../../Computation/State.h"
#include "../../Computation/DState.h"

// AmbientSpace Class Declaration

Geometry::Geometry()
{

}

Geometry::Geometry(
            std::function< Eigen::Matrix3d (Eigen::Vector3d)> metricTensor, 
            std::function< Eigen::Vector3d (State)> christoffel, 
            std::function<double(Eigen::Vector3d,Eigen::Vector3d)> distance
            )
{
    this->metricTensor = metricTensor;
    this->christoffel = christoffel;
    this->distance = distance;
}

Geometry::Geometry(
            std::function< Eigen::Matrix3d (Eigen::Vector3d)> metricTensor, 
            std::function< Eigen::Vector3d (State)> christoffel, 
            std::function<double(Eigen::Vector3d,Eigen::Vector3d)> distance, 
            derivative_funcs funcDict
            )
{
    this->metricTensor = metricTensor;
    this->christoffel = christoffel;
    this->distance = distance;
    this->funcDict = funcDict;
}

//Methods

DState Geometry::covariantAcceleration(State state)
{
    Eigen::Vector3d acc = this->christoffel(state);
    return DState(state.vel,acc);
}

double Geometry::dot(State state1, State state2)
{
    Eigen::Matrix3d mat = this->metricTensor(state1.pos);
    
    return state1.vel.transpose() * mat * state2.vel;
}

std::vector<State> Geometry::tangentBasis(Eigen::Vector3d pos)
{
    State b1 = State(pos, Eigen::Vector3d(1.,0.,0.));
    State b2 = State(pos, Eigen::Vector3d(0.,1.,0.));
    State b3 = State(pos, Eigen::Vector3d(0.,0.,1.));
    std::vector<State> basis = {b1,b2,b3};
    return basis;
}

State Geometry::gradient(std::function<double(Eigen::Vector3d)> fn, Eigen::Vector3d pos)
{
    std::vector<State> basis = tangentBasis(pos);
    State differential = State(pos, Eigen::Vector3d(0.,0.,0.));

    // #//add them all up:
    double df0 = basis[0].differentiate(fn);
    State b0 = basis[0].clone();
    b0.multiplyScalar(df0);
    differential.add(b0);

    double df1 = basis[1].differentiate(fn);
    State b1 = basis[1].clone();
    b1.multiplyScalar(df1);
    differential.add(b1);

    double df2 = basis[2].differentiate(fn);
    State b2 = basis[2].clone();
    b2.multiplyScalar(df2);
    differential.add(b2);

    // #//now the differential needs to be converted from a covector to a vector
    // #//using the inverse metric:
    Eigen::Matrix3d metric = this->metricTensor(pos);
    if(metric.determinant() < 0.00001)
    {
        std::cout << "Warning! Metric Tensor Near Singular" << std::endl;
    }

    Eigen::Matrix3d invMetric = metric.inverse();
    differential.vel = invMetric * differential.vel;

    return differential;
}

// This version overloaded to handle distance between vertices in configuration space class
State Geometry::gradient(std::function<double(Eigen::Vector3d,Eigen::Vector3d)> fn, Eigen::Vector3d posi, Eigen::Vector3d pos)
{
    std::vector<State> basis = tangentBasis(pos);
    State differential = State(pos, Eigen::Vector3d(0.,0.,0.));

    // #//add them all up:
    double df0 = basis[0].differentiate(fn, posi);
    State b0 = basis[0].clone();
    b0.multiplyScalar(df0);
    differential.add(b0);

    double df1 = basis[1].differentiate(fn, posi);
    State b1 = basis[1].clone();
    b1.multiplyScalar(df1);
    differential.add(b1);

    double df2 = basis[2].differentiate(fn, posi);
    State b2 = basis[2].clone();
    b2.multiplyScalar(df2);
    differential.add(b2);

    // #//now the differential needs to be converted from a covector to a vector
    // #//using the inverse metric:
    Eigen::Matrix3d metric = this->metricTensor(pos);
    if(metric.determinant() < 0.00001)
    {
        std::cout << "Warning! Metric Tensor Near Singular" << std::endl;
    }

    Eigen::Matrix3d invMetric = metric.inverse();
    differential.vel = invMetric * differential.vel;

    return differential;
}

